Exercise util templates

ocstl opened the pull request:
This is very much a WIP (mostly aimed at #698), but, before going any further down a potential rabbit hole, I though I'd ask for honest feedback. So, feel free to hurt my feelings. :smile: 

Essentially, this is an attempt to leverage `Deserialize` to simplify dealing with the canonical data for the tests file (see `structs::CanonicalData`), as well as (potentially) centralizing the logic in one place. Right now, the fields are private, with public getter methods, but I'm thinking that it might make more sense to make them publicly visible, as this is for internal use, and I don't think any invariants need to be maintained; but I might be wrong.

I've tried tinkering with the `syn` crate, but going through a `String` version of the functions seems the easiest path at this time, as they can relatively easily be turned into `syn::Item`s and, eventually, a `syn::File`.

The `structs::LabeledTest::generate_test_function` attempts some input unpacking through bindings that are then passed on to the relevant processing function, but this is very limited. JSON objects' keys being strings, this is fine when trying to unpack inputs, but is severely limiting if one needs to build a `HashMap` (or `BTreeMap`, etc.) with a different type for the keys. The same issue holds when trying to distinguish between `char`s and `String`s.

On that note, I think that trying to generate the processing functions first might be the way to go, as their signature can then guide the conversion, but the reverse might also be a good idea.

Anyway, as I said, any feedback would be most welcome at this stage.

coriolinus commented:
Just checking in. I apologize for not having responded to this earlier; I've been busy recently. I hope to have some time to look at this in the next few days, but I make no promises as to the date. Just wanted to let you know that this isn't being ignored; I've just got a bunch of non-volunteer stuff going on right now.

ocstl commented:
No apology needed. And take as much time as you need, I'm still trying to wrap my head around the `syn` crate at any rate. :smile:

ocstl commented on util/exercise/src/structs.rs:
```diff
@@ -0,0 +1,159 @@
+type Exercise = String;
+type Version = String;
+type Comments = Vec<String>;
+type TestGroup = Vec<LabeledTestItem>;
```
Thank you! I found that just going along with the schema made things (relatively) straightforward.

ocstl commented on util/exercise/src/structs.rs:
```diff
@@ -0,0 +1,159 @@
+impl LabeledTest {
+    pub fn generate_test_function(&self) -> String {
+        // Attributes.
+        let mut s = vec![String::from("#[test]"), String::from("#[ignore]")];
```
No, you didn't miss it. While this obviously sometimes adds an `#[ignore]` unnecessarily, I feel it will probably be easier to handle as a final step: find the first test function and drop the `#[ignore]` (or even just deleting the first `#[ignore]` of the file, since I believe only the test functions have them).

ocstl commented on util/exercise/src/structs.rs:
```diff
@@ -0,0 +1,159 @@
+}
+
+impl LabeledTest {
+    pub fn generate_test_function(&self) -> String {
```
The templates approach feels like a very good idea. Honestly, the `generate_test_function` method was about not breaking the previous behaviour; I should say mostly not breaking, as the unwrapping is a bit different (also `use_maplit`).

One interesting aspect with `syn` is that generating `ItemFn` is straightforward using `parse_str` (or even a full source `File`). And `ItemFn` implements `Eq` and `Hash`, which should make finding new, updated and deleted versions of the tests relatively straightforward.

That being said, I'm not very familiar with either `handlebars` or `tera`, and they may offer an even simpler way to compare old vs new than going through `syn`. Is that the case?

And they could probably deliver more readable code; `quote` is great, but, absent `rustfmt`, it does not deliver the most readable code from `syn`. :smile:

ocstl replied:
Looking back on this, implementing `generate_test_function` doesn't accomplish much; I had high hopes for the `Data` enum, but it hasn't lived up to it.

As it is, the final output is slightly different from what it used to be. So, this is one change I'm quite willing to revert to the original implementation in `lib.rs` (with the required changes to use `CanonicalData` instead). This would probably be for the best until we can wrestle the templates. What do you think?

coriolinus replied:
Yeah, I think that until we can get some proper templates in, I think I prefer the string-formatting "templates" over the imperative version. It just makes it easier, for me at least, to visualize what the output will be.

ocstl replied:
I agree. I reverted to the former `generate_test_function` function, although this required some minor modifications to use the public fields.

This still uses the `Data::to_literal` method, since it works fine, but I'm thinking this could also be reverted to the former `into_literal` function, by replacing the `input` and `expected` type with `Value`s; one modification would be to change the output of the function from a `Result<String>` to a `String`, since it can't fail as long as deserialization has succeeded (and eliminating the propagation to callers). In that case, I think resetting to an earlier commit (916d26b?) might make for a cleaner history though.

ocstl commented on util/exercise/src/structs.rs:
```diff
@@ -0,0 +1,159 @@
+type Optional = String;
+type Property = String;
+type Input = IndexMap<String, Data>;
+type Expected = Data;
```
I was tempted to do so because it feels weird to split them, but, on the other hand, this version has the advantage of shadowing very closely the JSON schema (although the `impl LabeledTest` from lines 64 to 104 muddies things a bit).

This is not a great reason to split them though, as it is less readable this way. I'll see how it looks without the macro calls and the intervening `impl`, but I think it will be better to group them.

ocstl replied:
I moved the `LabeledTest` implementation to after the structs and types. Having the types split up still feels a bit weird, but, combined with getting rid of the macro, it is easier to compare with the [JSON schema](https://github.com/exercism/problem-specifications/blob/master/canonical-schema.json). It made me notice that the `Expected` type might need some work to include an optional Error.

That being said, the structure is relatively simple, so grouping all the types shouldn't make maintenance/improvement very difficult.

ocstl commented on util/exercise/src/structs.rs:
```diff
@@ -0,0 +1,159 @@
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(untagged)]
+pub enum Data {
```
It started with a lot of good intentions. :laughing: 

It is interesting that we can deserialize into `i32`, `f64`, `char`, `String`, etc. The idea (hope) was that the `Map` variant would be a `IndexMap<Data, Data>`, which would give the type information; for example, for `alphametics`, the expected value as a `IndexMap<char, i32>` (and possibly `u8` eventually). It kinda works, but on a case-by-case basis (no pun intended); so we end up with a `IndexMap<char, i32>` for one test case and a `IndexMap<String, i32>` for another under the same property. Alas, my hopes were dashed. :sob: 

This is not to say that it's impossible, but I don't think asking `serde` to deliver this directly is possible. That being said, and thinking out loud, one possible way I've been mulling over is testing repeatedly `serde_json::value::Value`'s `is_` methods against all the inputs, and taking the "last" (or largest, etc.) type that fit all of them. As you mention, a newtype might help by implementing `PartialOrd` over the variants. But it does feel a bit hacky.

I do like the idea of a sub-enum, but I fear that it might encounter the issue related to case-by-case deserialization. That being said, it would probably be easier to implement the heuristic to find the best fit.

ocstl replied:
I've been thinking about this. Since `inputs` are always JSON objects, which we translate into `IndexMap`s, comparing keys across all the cases for a given property, using `HashSet<String>::symmetric_difference` for example, would tell us if we can use a struct or not. 

If they all share the same keys, and if we assume that for a the values for a given key all share the same type, unpacking into a struct is relatively straightforward. The assumption that the values are of the same type seems reasonable, but I have been burned in the past.

If they don't share the same keys, then a struct would require using `Option`s, and, at this point, a `HashMap` is probably the simpler choice. For example, in [`rest_api`](https://github.com/exercism/problem-specifications/blob/master/exercises/rest-api/canonical-data.json), the first case has no payload while the third one (the second one using `post`) does.

If we want to do this recursively, I think that using our own enum is probably the simplest solution (though maybe with a better name than `Data` :wink: ). It's not necessary an easy task, but I think it essentially boils down to using level order traversal.

coriolinus replied:
> If they all share the same keys, and if we assume that for a the values for a given key all share the same type, unpacking into a struct is relatively straightforward.

Agree, that seems a reasonable heuristic.

> If they don't share the same keys, [...] a HashMap is probably the simpler choice. 

Agree.

> If we want to do this recursively, ...

I'm not opposed to an iterative approach if one feels natural; it just seemed to me that a recursive traversal would be most natural, because we need to do type-detection and analysis on arbitrary sub-types. If (making this up) an exercise has `"input": {"a": {"b": "c"}}`, then we need to check that (bare minimum) all `input.a` also heuristically identify as structs (i.e. they all have a `b` whose value type is stringy) in order to categorize the whole `input` type as a struct instead of a map.

That said, any kind of type detection is going to end up being a vast improvement over the current state of affairs, which unconditionally uses maps, even when they're really not appropriate.

ocstl replied:
> I'm not opposed to an iterative approach if one feels natural; it just seemed to me that a recursive traversal would be most natural, because we need to do type-detection and analysis on arbitrary sub-types.

Agreed.

> If (making this up) an exercise has "input": {"a": {"b": "c"}}, then we need to check that (bare minimum) all input.a also heuristically identify as structs (i.e. they all have a b whose value type is stringy) in order to categorize the whole input type as a struct instead of a map.

I think I agree, but I'm not 100 % sure. Let's say `input.a` can't be reduced to a struct (sometimes the only key is `b`, sometimes `d`), so it has to "remain" a map. Couldn't we unpack it into a struct, `(Map<_, _>, String, String)` for example?

ocstl commented on util/exercise/src/structs.rs:
```diff
@@ -0,0 +1,159 @@
+                    .map(|(k, v)| format!("hm.insert(\"{}\", {});", k, v.to_literal()))
+                    .collect::<Vec<String>>()
+                    .join("\n");
+                format!("{{ let mut hm = HashMap::new();\n{}\n hm }}", data)
```
Good catch! I was toying with a `From<LabeledTest> for syn::ItemFn` implementation and couldn't think of a way to include it. But that's not a problem anymore. Thanks!

ocstl commented:
Sorry for the lack of recent activity. I haven't given up; I'm looking at the templates and trying to figure out a good way to proceed.

coriolinus commented:
No problem at all! I've spent the last several weeks consumed with family
and holidays, and have barely spent any time at a computer; I certainly
don't expect anyone else to have put in non-trivial volunteer time
programming at this time of year.

On Mon, Dec 30, 2019 at 10:55 PM Olivier Chassé St-Laurent <
notifications@github.com> wrote:

> Sorry for the lack of recent activity. I haven't given up; I'm looking at
> the templates and trying to figure out a good way to proceed.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/rust/pull/908?email_source=notifications&email_token=AB3V4TVANYAKRNRLY4Q3KULQ3JU4LA5CNFSM4JQXSRKKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH3JFUI#issuecomment-569807569>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB3V4TRSJSXKK42BPPUIBVDQ3JU4LANCNFSM4JQXSRKA>
> .
>

ocstl commented:
In the end, I reset to [916d26b](https://github.com/exercism/rust/commit/916d26bc7d02b3f32488dc59964ea8f06d45bbc2) to clean up the reverts, as the only "salvageable" part was using public fields instead of getter methods.

This is a very barebones implementation of templates. In the end, I opted for `Tera` instead of `handlebars` as it makes it easy to add the `use_maplit` option to the `Context` before rendering.

The formatting is pretty atrocious but `rustfmt` cleans it up very nicely when not using the `maplit` crate. I'm still somewhat wrapping my head around the way `Tera` handles formatting. Also, the snake casing of "description" and "property", `property | slugify | replace(from="-", to="_")`, leaves a bit to be desired to be fully honest, but this could be handled with a relatively simple function.

The issue with the first "#[ignore]" remains. This will be easier to handle at the test file level, which I held off implementing just yet (baby steps!); it **should** be straightforward when generating a new test file, but I'm not sure how easy this will be to handle when trying to update an existing one.

Anyway, any comment is most welcome.

coriolinus commented:
Cool! I don't have any personal preference between the template engines, so I'm happy to trust that Tera makes things easier than handlebars does. 

Agree that we see the `property | slugify | replace(...)` idiom often enough that it's probably worth a small dedicated function, even if the only savings is to change it to `property | snakeify` or whatever.

It would be interesting to see if the exercise util could automatically run rustfmt after generation; if so, that saves us a bunch of work trying to make the templates pretty.

Agree that omitting the first `#[ignore]` should be easy in generation. For updates, it's trickier, I agree; if we fully develop this tool, then the smartest solution I can think of works in phases:

1. Scan the existing file (AST?), build a map from function name to the literal text of the function body for all `fn test_*` and `fn process_*_case`.
2. Generate the new file as if from scratch.
3. Parse the new file, and merge. For all functions, if they appear only in the newly generated file, retain them. If they appear only in the old file, append them. If they appear in both, we handle `fn test_*` and `fn process_*_case` differently:
    - `fn test_*`: retain the new function, discard the old.
    - `fn process_*_case`: append the old function, discard the new.
4. We're almost certainly working with an AST at this point, so now sort all `process_*_case` before all `test_*` and alphabetize them. Leave the order of tests alone.

Still, that's a long way off, and probably outside the scope of this PR. This is great work so far!

ocstl commented:
Thanks for the review!

> Agree that we see the property | slugify | replace(...) idiom often enough that it's probably worth a small dedicated function, even if the only savings is to change it to property | snakeify or whatever.

I implemented it using closures around the "former" formatting functions, since they're still in use elsewhere in the code. This has the advantage of avoiding duplication and making it easier to translate between the rust code and the templates.

> It would be interesting to see if the exercise util could automatically run rustfmt after generation; if so, that saves us a bunch of work trying to make the templates pretty.

Agreed. The good news is that it already does when rustfmt is available. But rustfmt doesn't do formatting within the calls to the `hashmap!` macro. Which I can understand, as it is far from straightforward to parse it correctly.

> Agree that omitting the first #[ignore] should be easy in generation. For updates, it's trickier, I agree; if we fully develop this tool, then the smartest solution I can think of works in phases: [...]

I like the AST idea. The `syn` crate might be helpful, since the `ItemFn` struct implements `PartialEq`, so it becomes a matter of visiting every `ItemFn` in turn in both files. But this is a bit further in the future (and I tend to be optimistic about how easy things will be). Another PR feels right.

I'll finish up the template for a full test file, to make this "fully" functional (for some values of fully).

ocstl commented:
I added the template to generate a full test file, and modified "generate.rs" accordingly. The "dont_ignore" flag to avoid ignoring the first test feels a bit hacky, but it seems like the best solution to avoid adding complexity when generating a single test function. But I am very open to a better suggestion.

I left the handling of missing canonical data alone in "generate.rs". Given that the test file template relies on a for loop, it would probably be easy to generate a minimal `Context` to fill in the blanks (the "exercise" field), since the loop will ignore the undefined "cases". I am not adverse to making the required change, but this PR feels large enough as it is. :)

The format of the test file is slightly different, as it gathers all properties before generating the property functions at the end. This doesn't quite follow your suggestion in #686 (putting them before the tests), but it seemed the simplest way to at least keep them together. This can probably be more properly addressed with unpacking the inputs, as I expect this will require extracting the types across properties.

I'm still open to suggestions/improvements, but I'll switch this PR to "Ready for review". This has been quite a ride, thanks for your patience! :)

coriolinus commented on util/exercise/src/cmd/templates/test_file.rs:
```diff
@@ -0,0 +1,52 @@
+
+{# Prepare an array (global) to store the properties. Also, don't ignore the first case. -#}
+{% set properties = [] -%}
+{% set dont_ignore = true -%}
```
Maybe `first_test_case` would be clearer than `dont_ignore`, as a variable name?

ocstl replied:
Yes, yes it would. Thank you! :)

coriolinus commented on util/exercise/src/cmd/generate.rs:
```diff
@@ -78,50 +46,11 @@ fn generate_tests_from_canonical_data(
-    for property_body in property_functions.values() {
-        tests_file.write_all(property_body.as_bytes())?;
-    }
+    let mut context = Context::from_serialize(canonical_data)?;
```
While it's cool that you've managed to extract the set of properties within the template expansion, I do feel like the test file is much easier to read if the properties list comes first. It should be pretty straightforward to recursively traverse `canonical_data` at this point, discovering the proper set of properties, and then inject the properties list into the context.

ocstl replied:
I agree on the ordering. To be honest, I held off on implementing it because I expect that unpacking the inputs correctly will require grouping the test cases by property, to make sure all possible input formats are covered (for example, the rest_api exercise has a variable number of inputs).

But, we might as well do it now. :smile:

ocstl replied:
It was indeed pretty straightforward to implement in the end (see [afc5929](https://github.com/exercism/rust/pull/908/commits/afc592977fe4afc504b9451f00d5c95a44b81cfb)).

coriolinus commented on util/exercise/src/cmd/templates/test_fn.rs:
```diff
@@ -1,6 +1,9 @@
+{% import "macros.rs" as macros -%}
+
 #[test]
+{% if not dont_ignore -%}
```
I'm not familiar enough with Tera to know if this is possible, but could you set the global `dont_ignore` here, instead of up in `test_file.rs`? Like this:

```
{% if first_test_case -%}
{% set_global first_test_case = false -%}
{% else -%}
#[ignore]
{% endif -%}
```

coriolinus replied:
Part of the goal of this change suggestion would be to simplify case generation: for each case, there would only be a call to `{{ macros::gen_test_fn(case=item, dont_ignore=dont_ignore) }}`, because `dont_ignore` would be updated here, and the properties will be computed elsewhere.

ocstl replied:
That's a really good idea. Unfortunately, Keats/tera#336 mentions that macros can't interact with the global context, and I can confirm it after trying it out.

That being said, I really like your idea. Since the macro is the stumbling block, I'm thinking we could remove it and set up the variables "manually". This implies a bit of duplication (iterating over a group of cases vs a single case), but it's only a few lines and the duplicates remain next to each other in a relatively small file.

I haven't changed the name of the global variable yet, since the two changes would fit together nicely in a single commit.

coriolinus commented on util/exercise/src/cmd/templates/property_fn.rs:
```diff
@@ -0,0 +1,18 @@
+fn process_{{ format_property(property=property) }}_case<I, O>(input: I, expected: O) {
+    //  typical implementation:
+    //  assert_eq!(
+    //      student_{{ property | slugify | replace(from="-", to="_") }}_func(input),
```
```suggestion
    //      student_{{ format_property(property=property) }}_func(input),
```

ocstl replied:
Good catch, thanks!

coriolinus commented:
Note on merge type: squash and merge is the only legal way to merge to master right now. It's not what I prefer, but it's out of my hands.

ocstl commented:
Thank you for your help!

I'd also like to change "dont_ignore" to "first_test_case". Macros don't share the global Context, but it turns out included templates ({% include (...) %}) don't share the same global Context either.

coriolinus commented:
Sounds good! Good news: the PR to change the name can be _much_ smaller than this one.
